src/services/codefixes/fixAddMissingNewOperator.ts

```diff
@@ -0,0 +1,34 @@
+/* @internal */
+namespace ts.codefix {
+    const fixId = "addMissingNewOperator";
+    const errorCodes = [Diagnostics.Value_of_type_0_is_not_callable_Did_you_mean_to_include_new.code];
+    registerCodeFix({
+        errorCodes,
+        getCodeActions(context) {
+            const { sourceFile, span } = context;
+            const changes = textChanges.ChangeTracker.with(context, t => addMissingNewOperator(t, sourceFile, span));
+            return [createCodeFixAction(fixId, changes, Diagnostics.Add_missing_new_operator_to_call, fixId, Diagnostics.Add_missing_new_operator_to_all_calls)];
+        },
+        fixIds: [fixId],
+        getAllCodeActions: context => codeFixAll(context, errorCodes, (changes, diag) =>
+            addMissingNewOperator(changes, context.sourceFile, diag)),
+    });
+
+    function addMissingNewOperator(changes: textChanges.ChangeTracker, sourceFile: SourceFile, span: TextSpan): void {
+        const call = cast(findAncestorMatchingSpan(sourceFile, span), isCallExpression);
+        changes.insertNodeAt(sourceFile, span.start, createToken(SyntaxKind.NewKeyword), { suffix: " " });
+        if (isCallExpression(call.expression)) {

```

Comment:
I really feel this is masking a problem with the change tracker API. We shouldn't have to have special checks, the change tracker should do it automatically when replacing nodes, much like how our factories/printers automatically do this.

Reply:
Great example of where this won't work:

```ts
var foo;

foo()!()
```

Here, you'll end up with

```ts
new foo()!()
```

instead of

```ts
new (foo()!)()
```

We already have a `parenthesizeForNew` function in `factory.ts`: https://github.com/Microsoft/TypeScript/commit/30433c2c67cf8073b087e8ac90e36d1cf3a61c6b

So the change tracker should be doing something similar here.

Reply:
Ah, I see, that's a good point. My worry is that this won't be the last time we do something like this though. Maybe for now the right fix is to expose `parenthesizeForNew` in `compiler/utilities.ts` and use it here.

Reply:
Thanks for the feedback!

I'm trying to implement your suggestion but I don't think I fully understand the internals needed for it to work.

According to @andy-ms,  we should not use `replaceNode` and stick with the `insertNodeAt` so we won't generate the source file again.
Using the `insertNodeAt` doesn't really create a `new` node, but rather adds a `new` token to the start of the text span (without any change to the existing nodes or the text span).

On the other-hand, you are suggesting to use `parenthesizeForNew`, which receives an expression and returns it after wrapping it in parentheses if needed.

As I understand it, to work with the `parenthesizeForNew` API I'll need to create a `NewExpression` (similarly to what occurs in `createNew`), which brings us back to the `replaceNode` logic that we decided to avoid.

The only solution that aligns with both requirements is having a similar logic to the `switch` case in `parenthesizeForNew` in my code, that adds the parentheses (using `changes.insertNodeAt`) in these cases only (replacing the current `isCallExpression` check). 
But again, i'm not sure that this is what you guys are targeting for.

What do you think?